src/core/thread/mle.cpp

```diff
@@ -2202,7 +2202,7 @@ void Mle::ScheduleMessageTransmissionTimer(void)
     case kChildUpdateRequestActive:
 #if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
         // CSL transmitter may respond in next CSL cycle.
-        if (Get<Mac::Mac>().IsCslEnabled())
+        if (Get<Mac::Mac>().IsCslCapable() && !IsRxOnWhenIdle())
```

Comment:
This looks good to me. Just to make sure let me add my understanding to double-check. We have 
```c++
bool Mac::IsCslEnabled(void) const
{
    return !GetRxOnWhenIdle() && IsCslCapable();
}
```
But it won't be correct here, since we may not yet have set the `RxOnWhenIdle` on the `Mac` at this point, so need to check it in MLE directly. Is this correct?

Reply:
Your understanding is correct. I actually start to wonder if there is a better way to check if CSL is Enabled. Like e.g. getting a value of CSL period directly from radio driver.
Do You know why it was implemented as it is now?

Reply:
> Do You know why it was implemented as it is now?
Not really. I reviewed the PRs adding CSL, but was not myself directly involve din it is initial design.

If we see an opportunity to improve some aspects we should consider adding them.

How about changing the `IsCslEnabled()` to directly check with `Mle` on `RxOnWhenIdle` flag (instead of of `Mac`)? Can this work?

Reply:
@Irving-cl , thoughts?

Reply:
As I understand, `RxOnWhenIdle` in `Mac` indicates the Mac level RX behavior while that in `MeshForwarder` indicates the device mode in Thread level. I think `IsCslEnabled` is actually called from different contexts: Sometimes we want the condition returns true only when the Mac level is indeed RxOffWhenIdle (for example, in `Mac::UpdateIdleMode`). And sometimes, like this case, we want the condition returns true so long as the mode is set to `RxOffWhenIdle`. So my point is, we may not have a general `IsCslEnabled` and we still need to use different conditions in different cases. Thoughts? @abtink @canisLupus1313 

BTW, I don't quite understand how `RxOnWhenIdle` in `Mac` and `Meshforwarder` could be different in code above. `ScheduleMessageTransmissionTimer` is called by `SendChildUpdateRequest`. This should be done after Child attaches successfully. I think at that moment, `RxOnWhenIdle` in `Mac` is already set. IIRC, the Child ID Response is sent by CSL.

Reply:
@abtink 
> How about changing the IsCslEnabled() to directly check with Mle on RxOnWhenIdle flag (instead of of Mac)? Can this work?

In this use cases should be fine I not sure about the rest. This is why I haven't changed it so far. 

@Irving-cl 
> BTW, I don't quite understand how `RxOnWhenIdle` in `Mac` and `Meshforwarder` could be different in code above.

The certification testcase L5.3.6 does the transition form MED to SSED by change of `RxOnWhenIdle` state by firs setting it in `MLE` and then in `Meshforwarder`, `Mac`. There is a delay in synchronization of this flag and in between Openthread schedules the poll period based on those flags.

Either way I would propose to delegate this problem to some issue and continue discussion there.

Reply:
@canisLupus1313 

Make sense. We can discuss it somewhere else. Regarding `IsCslEnabled`, my point is that we should keep it as it is and uses `Get<Mac::Mac>().IsCslCapable() && !IsRxOnWhenIdle()` somewhere it's necessary. And probably we should add some comments to explain.

Reply:
@Irving-cl
I agree the comment is there already.